### PR TITLE
Fix importing from ES module

### PR DESCRIPTION
**Pull request by rchl**
This should fix the case when consumer of this package uses `type: 'module'` in `package.json`.

The issue with that was that in that case Node.js would resolve the import to the UMD bundle and then importing with `import { URI } from 'vscode-uri'` would not work. Only a hacky way like `import vscodeUri from 'vscode-uri'; vscodeUri.URI(...);` would work.

I can make an example repo to demonstrate the issue if you want.

(Note that the `module` field in `package.json` is ignored by Node.js so that's why I've added `exports`. The `module` field is only supported by bundlers like Webpack.)

BTW. Please sort out this repo first -- `main` is lagging behind as there is a 3.0.3 tag but changes from it are not merged to it.

(And this repo could have issues enabled as there are some type of issues like this one that don't belong in vscode repo itself.)

### CODE REVIEW

**Review by remcohaszing — APPROVED** [review]

**Comment by rchl** [comment]
@dbaeumer any comments? :)

**Comment by dbaeumer** [comment]
That's more @jrieken

File: package.json [hunk 1]
```diff
@@ -4,6 +4,10 @@
   "version": "3.0.2",
   "description": "The URI implementation that is used by VS Code and its extensions",
   "main": "./lib/umd/index.js",
+  "exports": {
+    "import": "./lib/esm/index.mjs",
+    "require": "./lib/umd/index.js"
```
Comment (remcohaszing) [inline package.json]: This also needs `exports.types` for the `node16` / `nodenext` module resolution in the TypeScript compiler options.

```suggestion
    "require": "./lib/umd/index.js",
    "types": "./lib/umd/index.d.ts"
```

Apart from that, great addition!

**Review by remcohaszing — CHANGES_REQUESTED** [review]

File: package.json (hunk 1 above)
Comment (rchl) [reply package.json]: I believe that I didn't need that when testing with https://github.com/typescript-language-server/typescript-language-server/ but I will do some more testing later. I'm already using `"moduleResolution": "Node16"` there.

Also I wonder whether "exports.types" is specific to only some of the following: Node, Webpack, Typescript...
I thought that `typescript` followed the root `types`.

Comment (remcohaszing) [reply package.json]: You can try it by creating the following files:

`package.json`

```json
{
  "type": "module",
  "dependencies": {
    "typescript": "^4.8.2",
    "vscode-uri": "^3.0.3"
  }
}
```

`tsconfig.json`

```jsonc
{
  "compilerOptions": {
    // implies "moduleResolution": "node16",
    "module": "node16",
    "noEmit": true,
    "strict": true,
  }
}
```

`foo.ts`

```ts
import {URI} from 'vscode-uri'

console.log(URI.parse('file:///foo.ts'))
```

Then make the modifications in `node_modules/vscode-uri/package.json`. To test, simply run `npx tsc` or `yarn tsc`. You’ll find out `exports.types` is indeed required.

Unfortunately a lot of existing packages aren’t properly configured yet. I’m glad to see someone making this effort. :smile:

Comment (rchl) [reply package.json]: Thanks. Tested and confirmed your findings.

**Comment by rchl** [comment]
Any interest in getting this?

It will likely be a breaking change for consumers with `type: module` set so might require major version to be fully proper.

**Comment by jrieken** [comment]
@rchl VS Code extension cannot really use ESM (yet) so I wonder where/how you would use this?

**Comment by remcohaszing** [comment]
Some language servers / services use it. Also VSCode extensions are often bundled using esbuild. In this case CJS will result in boilerplate code in the bundle to provide CJS interoperability. In most cases this could be replaced with WHATWG [URL](https://developer.mozilla.org/en-US/docs/Web/API/URL) though.

I think the question shouldn’t be where / how this would be used though. This package already contains ESM. If an npm package is published containing ESM, IMO it should contain real ESM, not faux. That’s what this PR is about.

**Comment by rchl** [comment]
I guess I'll reiterate most of the points that @remcohaszing made already but I'm using it in https://github.com/typescript-language-server/typescript-language-server/ and as @remcohaszing pointed out, this package is already set up with ES bundle (the  `"module": "./lib/esm/index.js"` property in `package.json`). I'm just adding extra properties so that the ES bundle also works correctly with native Node ES modules support rather than only with bundlers which tend to be more forgiving when it comes to handing incorrect exports.

**Comment by rchl** [comment]
And to clarify more: this change doesn't make it a "ES modules only" package which would be the case if it had a `"type": "module"` property set in `package.json`. It just makes it compatible with consumers of this package that use `"type": "module"`.

**Comment by Jason3S** [comment]
@jrieken,

Any progress here?

The bug is that `package.json` contains `"module": "./lib/esm/index.js",`. Removing that line would allow it to work when used with a `.mjs` files.

**Comment by aeschli** [comment]
Fixed by https://github.com/microsoft/vscode-uri/pull/39

**Pull request closed without merging**